engine/execution/state/state.go

```diff
@@ -50,6 +51,24 @@ type ReadOnlyExecutionState interface {
 	GetCollection(identifier flow.Identifier) (*flow.Collection, error)
 }
 
+// IsBlockExecuted returns whether the block has been executed.
+// it checks whether the statecommitment exists in execution state.
+func IsBlockExecuted(ctx context.Context, state ReadOnlyExecutionState, block flow.Identifier) (bool, error) {
+	_, err := state.StateCommitmentByBlockID(ctx, block)
```

Comment:
This is the only part I'm worried about, the block might be executed and state commitment might be there but it doesn't provide any guarantees that Result has been stored, which causes this issue: 
 - if all exec nodes crash at the same time (after state commitment storage and before result storage), consensus nodes will miss a result forever and sealing will halt.

Reply:
while we are waiting for the other issue (transactional updates) to happen I think we should check the exec results (or any thing that is the last piece we store) instead of state commitments and make sure we can re-execute an already executed block. an alternative is to always start re-executing the last executed block anyway.

Reply:
Agreed. I implemented [the alternative](https://github.com/onflow/flow-go/pull/73/files#diff-e5c0dea70cf688717060471583da13cc5269a6f661e955613ed7ef698354cc1aR292-R318)